diesel/src/query_builder/update_statement/changeset.rs

```diff
@@ -68,20 +69,95 @@ where
 }
 
 #[derive(Debug, Clone, Copy, QueryId)]
-pub struct Assign<Col, Expr> {
-    _column: Col,
+pub struct Assign<Target, Expr> {
+    target: Target,
     expr: Expr,
 }
 
 impl<T, U, DB> QueryFragment<DB> for Assign<T, U>
 where
     DB: Backend,
-    T: Column,
+    T: QueryFragment<DB>,
     U: QueryFragment<DB>,
 {
     fn walk_ast<'b>(&'b self, mut out: AstPass<'_, 'b, DB>) -> QueryResult<()> {
-        out.push_identifier(T::NAME)?;
+        QueryFragment::walk_ast(&self.target, out.reborrow())?;
         out.push_sql(" = ");
-        QueryFragment::walk_ast(&self.expr, out)
+        QueryFragment::walk_ast(&self.expr, out.reborrow())
+    }
+}
+
+/// Represents the left hand side of an assignment expression for an
+/// assignment in [AsChangeset]. The vast majority of the time, this will
+/// be a [Column]. However, in certain database backends, it's possible to
+/// assign to an expression. For example, in Postgres, it's possible to
+/// "UPDATE TABLE SET array_column\[1\] = 'foo'".
+///
+/// You shouldn't have to implement this trait. Expressions can be made
+/// assignable with [AssignableExpression].
+pub trait AssignmentTarget {
+    /// Table the assignment is to
+    type Table: Table;
+    /// A wrapper around a type to assign to (this wrapper should implement
+    /// [QueryFragment]).
+    type QueryAstNode;
+
+    /// Move this in to the AST node which should implement [QueryFragment].
+    fn into_target(self) -> Self::QueryAstNode;
+}
```

Comment:
This likely should be defined in a `mod private {}` + then be rexported as `pub(crate)` item from this module to make this diesel private.

See here: https://github.com/diesel-rs/diesel/blob/c237a74b63900b734c659a201fc1d6b87108033b/diesel/src/backend.rs#L390 for an example of this strategy.